py/server/tests/test_barrage.py

```diff
@@ -0,0 +1,130 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+import os
+import unittest
+
+import subprocess
+import time
+
+from deephaven import DHError
+from deephaven.barrage import barrage_session
+
+from tests.testbase import BaseTestCase
+
+
+class BarrageTestCase(BaseTestCase):
+    shared_ticket = None
+    server_proc = None
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        env = {"START_OPTS": "-DAuthHandlers=io.deephaven.auth.AnonymousAuthenticationHandler"}
+        env.update(dict(os.environ))
+        cls.server_proc = subprocess.Popen(["/opt/deephaven/server/bin/start"], shell=False, env=env,
+                                            stdin=subprocess.PIPE,
+                                            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        cls.ensure_server_running()
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.server_proc.kill()
+        super().tearDownClass()
+
+    @classmethod
+    def ensure_server_running(cls):
+        from pydeephaven.session import Session, SharedTicket
+
+        for _ in range(30):
+            try:
+                Session("localhost", 10000)
+                break
+            except Exception as e:
+                time.sleep(1)
+        else:
+            raise RuntimeError("Cannot connect to the server")
+
+    def setUp(self) -> None:
+        from pydeephaven.session import Session, SharedTicket
+
+        self.pub_session = Session("localhost", 10000)
+        self.t = self.pub_session.empty_table(1000).update(["X = i", "Y = 2*i"])
+        self.shared_ticket = SharedTicket.random_ticket()
+        self.pub_session.publish_table(self.shared_ticket, self.t)
+
+    def tearDown(self) -> None:
+        self.pub_session.close()
+
+    def test_barrage_session(self):
+        session = barrage_session(host="localhost", port=10000, auth_type="Anonymous")
+        self.assertIsNotNone(session)
+
+        with self.assertRaises(DHError):
+            barrage_session(host="invalid", port=10000, auth_token="Anonymous")
+
+        with self.assertRaises(DHError):
+            barrage_session(host="localhost", port=10000, auth_type="Basic", auth_token="user:password")
+
+    def test_subscribe(self):
+        session = barrage_session(host="localhost", port=10000, auth_type="Anonymous")
+        t = session.subscribe(ticket=self.shared_ticket.bytes)
+        self.assertEqual(t.size, 1000)
+        self.assertEqual(len(t.columns), 2)
+        sp = t.snapshot()
+        self.assertEqual(sp.size, 1000)
+        t1 = t.update("Z = X + Y")
+        self.assertEqual(t1.size, 1000)
+
+        with self.subTest("using barrage session as a context manager"):
+            with barrage_session(host="localhost", port=10000, auth_type="Anonymous") as cm:
+                t = cm.subscribe(ticket=self.shared_ticket.bytes)
+            t1 = t.update("Z = X + Y")
```

Comment:
Is `t` valid outside of the CM even though the session has been closed?  Does it have data but not tick?

Reply:
Yes and yes.

Reply:
We should check in w/ @rcaudy

Reply:
My answer was based on the test you did yesterday.

Reply:
The behavior with subscribed tickets after the subscribing session is close is changed. They becomes invalid after the session is closed.